### ISSUE TITLE
Renderer rewrite

### ISSUE THREAD

**Issue by xycaleth** [JACoders/OpenJK#54]
I propose that the renderer is rewritten or at least updated to take advantage of the newer graphics techniques which have been available for the past 10 years or so (they were introduced around the same time that JKA was released).

There are many ways the current renderer can be improved. The bottlenecks in the current renderer (from an educated guess) are:
- All vertex processing is performed on the CPU. All games nowadays offload this work to the GPU which is designed to handle these types of calculations (in parallel).
- All vertex data must be transferred to the graphics card every frame. Using vertex buffers allows the vertex data to remain resident on the graphics card which are referenced by identifiers. Instead of sending all the vertex to render, the renderer can instead tell the GPU to render vertices from a specific vertex buffer instead.
- Each shader stage is a separate 'render pass'. This means that a surface may have to rendered multiple times in order to generate the desired outcome. I half-wrote a Q3 shader to GLSL shader translator a few years ago which I may consider finishing to help with this. By doing this, most cases the shader stages should be able to be merged into one render pass.

I'll post a proposed renderer design some point in the near future (hopefully). I probably won't have enough time to code it until June/July so by posting the design, there's a chance someone else can start work on it.
